doc/authorization_policy_engine.md

```diff
@@ -0,0 +1,235 @@
+| req           | The API call request body (not available on client or bidirectional streaming RPC calls) | { "filter": {} } |
+
+The request (`req`) is the marshalled JSON object from the [SPIRE
+api sdk](https://github.com/spiffe/spire-api-sdk/). Note that it is not
+available on client or bidirectional streaming RPC API calls.
+
+## Permissions file (databinding)
+
+The permissions file consists of a JSON blob which represents the data that is
+used in the evaluation of the policy. This is generally free-form and can be
+used in the rego policy in any way. Data in this JSON blob is pre-compiled into
+the policy evaluation on the policy engine evaluation. Therefore, there it is
+recommended to put as much data as possible in the databinding so that it can be
+optimized by the policy engine.
+
+These data objects can be accessed via the `data` field in the rego policy. For
+example, a JSON data object may look like this:
+
+```
+{
+    "apis": [
+        { "full_method": "/spire.api.server.svid.v1.SVID/MintJWTSVID" },
+        { "full_method": "/spire.api.server.bundle.v1.Bundle/GetFederatedBundle"},
+        { "full_method": "/spire.api.server.svid.v1.SVID/BatchNewX509SVID"}
+    ]
+}
+```
+
+With the example data object above, we could construct a policy in rego to 
+check that if the input's full method is equal to one of the objects defined in 
+the `apis` fields' `full_method` sub-field, then `allow` should be set to true.
+```
+allow = true {
+    input.full_method == data.apis[_].full_method
+}
+```
+
+### Default configurations
+
+Here are the default set of rego policy and permissions/databindings. These are
+what is required to carry out the default authorization decisions as highlighted
+at the start of this document.
+
+#### Default policy.rego
+
+The default rego policy is located [here](/pkg/server/authpolicy/policy.rego).
+
+#### Default permissions.json (databindings)
+
+The default permissions.json is located [here](/pkg/server/authpolicy/permissions.json).
+
+The default permissions file contains a field called "apis".
+This field has a list of APIs that is current being configured with the rego
+policy.
+
+The fields of each object are as follows:
+
+| field             | Description | Example |
+| ----------------- | ----------- | ------- |
+| full_method       | The full method name of the API call | /spire.api.server.svid.v1.SVID/MintJWTSVID |
+| allow_any         | if true, sets result.allow to true | |
+| allow_local       | if true, sets result.allow_if_local to true | |
+| allow_admin       | if true, sets result.allow_if_admin to true | |
+| allow_downstream  | if true, sets result.allow_if_downstream to true | |
+| allow_agent       | if true, sets result.allow_if_agent to true | |
+
+# Extending the policy
+
+This section contains examples of how the authorization policy can be extended.
+
+## Example 1: Entry creation namespacing restrictions
```

Comment:
I've tried to follow this example. I've built the server and agent using `make build` after editing `policy.rego` and `permissions.json`. 

As a test, I've done the following:
*  I've created an agent with the following SPIFFE ID:  **spiffe://example.org/schedulers/finance**
* I then created the following entry
  + SPIFFE ID **//example.org/schedulers/hr**
   + Parent ID **spiffe://example.org/schedulers/finance**

I managed to create the entry and even receive the corresponding SVID from the Workload API (via [ghostunnel](https://github.com/ghostunnel/ghostunnel)).
Shouldn't this have failed?

Reply:
Hey Benardi! Let's chat on slack see whether we can figure out what the issue is - or whether my OPA code has a bug

Reply:
Sure thing.

Reply:
So far, we've established that as I was using `spire-server entry create`, the call to BatchCreateEntry API must have been marked as local, and thus namespacing was bypassed.

I still have to surmise a way to make a non-local call to the BatchCreateEntry API to discern whether the namespacing will be enforced.

Reply:
Hey Benardi,
Did you find a way to test calling the Entry API through Ghostunnel?
I'm looking to test something simliar. I've got the [tunnel](https://github.com/ghostunnel/ghostunnel/tree/master/docs/spiffe-workload-api-demo/unix) working was going to configure the front-end-agent to call the server tcp connection, but unsure how to form the request.
Many thanks
Phil